src/main/java/se/liu/ida/hefquin/engine/data/mappings/impl/EntityMappingImpl.java

```diff
@@ -25,26 +30,103 @@ public EntityMappingImpl( final Graph mappingDescription ) {
+				locals.add(workingTriple.getObject());
+			} else {
+				// g2lmap doesn't contain any local URI for this global yet
+				final Set<Node> localSet = new HashSet<>();
+				g2lMap.put( workingTriple.getSubject(), localSet );
+				localSet.add(workingTriple.getObject());
+			}
+			
+			// The same for l2g.
+			final Set<Node> globals = l2gMap.get(workingTriple.getObject());
+			if(globals != null) {
+				globals.add(workingTriple.getSubject());
+			} else {
+				final Set<Node> newGlobalSet = new HashSet<>();
+				l2gMap.put( workingTriple.getObject(), newGlobalSet );
+				newGlobalSet.add(workingTriple.getSubject());
+			}
 		}
 	}
 
 	@Override
 	public Set<TriplePattern> applyToTriplePattern( final TriplePattern tp ) {
-		// TODO: implement this function
-		return Collections.singleton(tp);
+		final Set<Triple> workingSet = new HashSet<Triple>();
+		if(tp.asJenaTriple().getSubject().isURI()) {
+			final Set<Node> localSubjects = g2lMap.get( tp.asJenaTriple().getSubject() );
+			if( localSubjects != null ) {
+				for (final Node localSubject : localSubjects) {
+					final Triple workingTriple = new Triple(localSubject,tp.asJenaTriple().getPredicate(), tp.asJenaTriple().getObject());
+					workingSet.add(workingTriple);
+				}
+			} else {
+				workingSet.add(tp.asJenaTriple());
+			}
+		} else {
+			workingSet.add(tp.asJenaTriple());
+		}
+		final Set<TriplePattern> resultSet = new HashSet<TriplePattern>();
+			if(tp.asJenaTriple().getObject().isURI()) {
+			final Set<Node> localObjects = g2lMap.get(tp.asJenaTriple().getObject());
+			if(g2lMap.containsKey(tp.asJenaTriple().getObject())) {
+				for (final Triple workingTriple : workingSet) {
+					for (final Node localObject : localObjects) {
+						final Triple resultTriple = new Triple(workingTriple.getSubject(), workingTriple.getPredicate(), localObject);
+						resultSet.add(new TriplePatternImpl(resultTriple));
+					}
+				}
+				return resultSet;
+			} else {
+				for (final Triple workingTriple : workingSet ) {
+					resultSet.add(new TriplePatternImpl(workingTriple));
+				}
+				return resultSet;
+			}
+		} else {
+			for (final Triple workingTriple : workingSet ) {
+				resultSet.add(new TriplePatternImpl(workingTriple));
+			}
+			return resultSet;
+		}
 	}
 
 	@Override
 	public Set<SolutionMapping> applyToSolutionMapping( final SolutionMapping sm ) {
-		// TODO: implement this function
-		return Collections.singleton(sm);
+		final Set<SolutionMapping> newMappings = new HashSet<SolutionMapping>();
+		final Binding binding = sm.asJenaBinding();
+		final Iterator<Var> it = binding.vars();
+		while (it.hasNext()) {
+			final Var var = it.next();
+			final Node node = binding.get(var);
+			final Set<Node> mappedNodes = g2lMap.get(node);
+			if (mappedNodes == null) { // Local Node not different from global. 
+				final Binding newBinding = BindingFactory.binding(var,node);
+				final SolutionMapping newMapping = new SolutionMappingImpl(newBinding);
+				newMappings.add(newMapping);
```

Comment:
That's incorrect because it considers only one of the variables in the given solution mapping. Even worse, it considers each variable in the given solution mapping separately.

Consider the following example: Assume the given solution mapping `sm` is: sm = { ?x -> ex:Bob, ?y -> ex:Ali }.

If neither ex:Bob nor ex:Ali are mentioned in the entity mapping, then the `applyToSolutionMapping` function should return `sm` unchanged. Instead, your current implementation returns two new solution mappings, one of them only with ?x -> ex:Bob and the other one only with ?y -> ex:Ali.

If, in contrast, the entity mapping maps ex:Bob to ex:Robert and ex:Ali to ex:Alibaba, then the `applyToSolutionMapping` function should return a single new solution mapping with ?x -> ex:Robert and ?y -> ex:Alibaba. Instead, your current implementation (in the else-branch below) creates two separate solution mappings, one of them only with ?x -> ex:Robert and the other one only with ?y -> ex:Alibaba.

If the entity mapping does not map ex:Ali, but it maps ex:Bob to both ex:Robert and ex:Bobby, then the `applyToSolutionMapping` function should return two new solution mappings, one with ?x -> ex:Robert and ?y -> ex:Ali, and the other one with ?x -> ex:Bobby and ?y -> ex:Ali. Instead, your current implementation would return three solution mappings.